### ISSUE TITLE
build on RHEL 7 against available EPEL packages

### ISSUE THREAD

**Issue by garlick** [flux-framework/distribution#4]
It would be easy on our users if our flux distribution built on top of EPEL packages on RedHat based distros.  [EPEL 7](https://dl.fedoraproject.org/pub/epel/7/x86_64/repoview/)  provides
- zeromq-4.0.5
- czmq-1.4.1
- openpgm-5.2.122
- libsodium-1.0.5

A few issues:
- This is an ancient czmq (current is 3.0.2).
- Unsure if zeormq is built with libsodium and openpgm; if not we would need to handle missing crypto and pgm better.

**Comment by grondo** [comment]
Checking on an RHEL7 system, python-2.7 is now included, but python-cffi-0.8.6 seems to be the only version built for EPEL7 (and seems to be required by python-cryptography:
```
$ rpm -q --whatrequires python-cffi
python-cryptography-0.8.2-1.el7.x86_64
```

I don't know enough to know if we'll be able to update the python-cffi version. 

@trws: Is there any way to support python-cffi-0.8.6 in flux-core?

**Comment by grondo** [comment]
FYI upstream koji builds for python-cffi are found here:

 http://koji.fedoraproject.org/koji/packageinfo?packageID=16680

Looks like they updated python-cffi in fedora 23/24

**Comment by trws** [comment]
In theory it could be done, but it would be a non-trivial amount of work, because 1.0 is where it became possible to generate the bindings ahead of time and build them into a library the way it does right now.  To support a version lower than 1.0, all of the filtered header code would have to be installed with the python bindings and processed and re-compiled on every execution of a script using the bindings.  I, at least, would very much prefer to avoid this if possible.


________________________________

**Comment by grondo** [comment]
@trws, Thanks! That doesn't sound worth it so we'll see if we can't update the EPEL version. If cffi is backward compatible I don't think it should be too big a deal.

**Comment by garlick** [comment]
I should have mentioned a few comments ago that I built the latest zeromq and czmq packages in our RHEL 7 koji build farm, and that @grondo asked our Red Hat guy to work with EPEL 7 "upstream" to get the EPEL packages updated and built with crypto so flux can just work on RHEL/CentOS 7.  The current EPEL czmq is ridiculously old, and zeromq is built without crypto, hence unusable for Flux.

**Comment by garlick** [comment]
Here's a link to the RH bugs opened by our guy:

[1301163: zeromq-4.1.4 is available](https://bugzilla.redhat.com/show_bug.cgi?id=1301163)
[1301197:  RFE add libgssapi_krb5 to zeromq 4.1.4](https://bugzilla.redhat.com/show_bug.cgi?id=1301197)

**Comment by grondo** [comment]
Should we also ask if czmq can be upgraded?

**Comment by garlick** [comment]
Yeah, sorry, I thought that was already in Ben's queue.

**Comment by trws** [comment]
We talked about pulling in judy as well, maybe put that on the list while we're thinking about it? I'm not sure if it's packaged, but it hasn't been updated in quite a while, so I imagine whatever's in there would be current.

Side-note by the way, I ran into a nice c unit test library the other day, called cmocka, that gives a number of the features of the fancy C++ ones with nothing but the C standard library.  It also has built-in TAP output... I'm toying with it for flux-related side code, might be a nice way to structure things like death tests (it handles and passes signals and other non-abort deaths automatically).

**Comment by garlick** [comment]
Looks like Ben opened this:

[1302118: Please update czmq to version 3.0.2](https://bugzilla.redhat.com/show_bug.cgi?id=1302118)

**Comment by garlick** [comment]
> We talked about pulling in judy as well, maybe put that on the list while we're thinking about it? I'm not sure if it's packaged, but it hasn't been updated in quite a while, so I imagine whatever's in there would be current.

It seems to be a part of EPEL 7 which makes it easy to pull in to our TOSS image.  I'll open a ticket for our admins to do that.
https://dl.fedoraproject.org/pub/epel/7/x86_64/j/Judy-1.0.5-8.el7.x86_64.rpm
https://dl.fedoraproject.org/pub/epel/7/x86_64/j/Judy-devel-1.0.5-8.el7.x86_64.rpm

**Comment by garlick** [comment]
Ticket opened:
[TOSS-3071: include Judy in TOSS 3 image](https://lc.llnl.gov/jira/browse/TOSS-3071)

**Comment by garlick** [comment]
Just tagged 0.1.0 and uploaded a make dist tarball:
https://github.com/flux-framework/flux-core/files/107523/flux-core-0.1.0.tar.gz

**Comment by garlick** [comment]
I'm currently blocked on the python-cffi requirement.  There is this:

[1297585: Rebase python-cffi to 1.4.2](https://bugzilla.redhat.com/show_bug.cgi?id=1297585)

**Comment by garlick** [comment]
I'll see if I can push the newer python-cffi and python-crpytography through koji this morning.  It seems relatively safe to replace those packages in TOSS 3 if RHEL 7.4(?) is going that way...

**Comment by garlick** [comment]
I produced a scratch build for TOSS3 minus python (python-cffi still hasn't been built/updated).  Here's the manifest in case people have any issues with the locations of things
```
/etc/ld.so.conf.d/flux-libs.conf
/etc/wreck
/etc/wreck/lua.d
/etc/wreck/lua.d/01-env.lua
/etc/wreck/lua.d/02-affinity.lua
/etc/wreck/lua.d/input.lua
/etc/wreck/lua.d/output.lua
/etc/wreck/lua.d/timeout.lua
/usr/bin/flux
/usr/include/flux
/usr/include/flux/core
/usr/include/flux/core.h
/usr/include/flux/core/attr.h
/usr/include/flux/core/barrier.h
/usr/include/flux/core/conf.h
/usr/include/flux/core/connector.h
/usr/include/flux/core/dispatch.h
/usr/include/flux/core/event.h
/usr/include/flux/core/flog.h
/usr/include/flux/core/flux.h
/usr/include/flux/core/handle.h
/usr/include/flux/core/heartbeat.h
/usr/include/flux/core/info.h
/usr/include/flux/core/jstatctl.h
/usr/include/flux/core/kvs.h
/usr/include/flux/core/live.h
/usr/include/flux/core/message.h
/usr/include/flux/core/modctl.h
/usr/include/flux/core/module.h
/usr/include/flux/core/mrpc.h
/usr/include/flux/core/panic.h
/usr/include/flux/core/reactor.h
/usr/include/flux/core/reduce.h
/usr/include/flux/core/reparent.h
/usr/include/flux/core/request.h
/usr/include/flux/core/response.h
/usr/include/flux/core/rpc.h
/usr/include/flux/core/security.h
/usr/include/flux/pmi.h
/usr/include/flux/pmi2.h
/usr/include/flux/pmi_deprecated.h
/usr/lib64/flux
/usr/lib64/flux-core/build
[snip - entire source tree here, with only libtap built]
/usr/lib64/flux/connectors
/usr/lib64/flux/connectors/local.so
/usr/lib64/flux/connectors/loop.so
/usr/lib64/flux/connectors/shmem.so
/usr/lib64/flux/libbarrier.so
/usr/lib64/flux/libbarrier.so.0
/usr/lib64/flux/libbarrier.so.0.0.0
/usr/lib64/flux/libflux-core.so
/usr/lib64/flux/libflux-core.so.0
/usr/lib64/flux/libflux-core.so.0.0.0
/usr/lib64/flux/libjsc.so
/usr/lib64/flux/libjsc.so.0
/usr/lib64/flux/libjsc.so.0.0.0
/usr/lib64/flux/libkvs.so
/usr/lib64/flux/libkvs.so.0
/usr/lib64/flux/libkvs.so.0.0.0
/usr/lib64/flux/liblive.so
/usr/lib64/flux/liblive.so.0
/usr/lib64/flux/liblive.so.0.0.0
/usr/lib64/flux/libmodctl.so
/usr/lib64/flux/libmodctl.so.0
/usr/lib64/flux/libmodctl.so.0.0.0
/usr/lib64/flux/libmrpc.so
/usr/lib64/flux/libmrpc.so.0
/usr/lib64/flux/libmrpc.so.0.0.0
/usr/lib64/flux/libpmi.so
/usr/lib64/flux/libpmi.so.0
/usr/lib64/flux/libpmi.so.0.0.0
/usr/lib64/flux/modules
/usr/lib64/flux/modules/barrier.so
/usr/lib64/flux/modules/connector-local.so
/usr/lib64/flux/modules/content-sophia.so
/usr/lib64/flux/modules/content-sqlite.so
/usr/lib64/flux/modules/job.so
/usr/lib64/flux/modules/kvs.so
/usr/lib64/flux/modules/live.so
/usr/lib64/flux/modules/mecho.so
/usr/lib64/flux/modules/modctl.so
/usr/lib64/flux/modules/resource-hwloc.so
/usr/lib64/flux/modules/wrexec.so
/usr/lib64/lua/5.1/flux
/usr/lib64/lua/5.1/flux.so
/usr/lib64/lua/5.1/flux/affinity.so
/usr/lib64/lua/5.1/flux/hostlist.so
/usr/lib64/pkgconfig/flux-core.pc
/usr/libexec/flux
/usr/libexec/flux/core
/usr/libexec/flux/core/flux-comms
/usr/libexec/flux/core/flux-comms-stats
/usr/libexec/flux/core/flux-config
/usr/libexec/flux/core/flux-event
/usr/libexec/flux/core/flux-exec
/usr/libexec/flux/core/flux-hwloc
/usr/libexec/flux/core/flux-jstat
/usr/libexec/flux/core/flux-keygen
/usr/libexec/flux/core/flux-kvs
/usr/libexec/flux/core/flux-logger
/usr/libexec/flux/core/flux-lstopo
/usr/libexec/flux/core/flux-module
/usr/libexec/flux/core/flux-mping
/usr/libexec/flux/core/flux-ping
/usr/libexec/flux/core/flux-ps
/usr/libexec/flux/core/flux-snoop
/usr/libexec/flux/core/flux-start
/usr/libexec/flux/core/flux-submit
/usr/libexec/flux/core/flux-topo
/usr/libexec/flux/core/flux-up
/usr/libexec/flux/core/flux-wreck
/usr/libexec/flux/core/flux-wreckrun
/usr/libexec/flux/flux-broker
/usr/libexec/flux/wrexecd
/usr/share/flux
/usr/share/flux/help.d
/usr/share/flux/help.d/core.json
/usr/share/lua/5.1/flux
/usr/share/lua/5.1/flux/alt_getopt.lua
/usr/share/lua/5.1/flux/base64.lua
/usr/share/lua/5.1/flux/lustache
/usr/share/lua/5.1/flux/lustache.lua
/usr/share/lua/5.1/flux/lustache/context.lua
/usr/share/lua/5.1/flux/lustache/renderer.lua
/usr/share/lua/5.1/flux/lustache/scanner.lua
/usr/share/lua/5.1/flux/posix.lua
/usr/share/lua/5.1/flux/timer.lua
/usr/share/lua/5.1/wreck
/usr/share/lua/5.1/wreck.lua
/usr/share/lua/5.1/wreck/io.lua
/usr/share/man/man1/flux-broker.1.gz
/usr/share/man/man1/flux-config.1.gz
/usr/share/man/man1/flux-content.1.gz
/usr/share/man/man1/flux-dmesg.1.gz
/usr/share/man/man1/flux-env.1.gz
/usr/share/man/man1/flux-exec.1.gz
/usr/share/man/man1/flux-getattr.1.gz
/usr/share/man/man1/flux-keygen.1.gz
/usr/share/man/man1/flux-kvs.1.gz
/usr/share/man/man1/flux-list-instances.1.gz
/usr/share/man/man1/flux-logger.1.gz
/usr/share/man/man1/flux-lsattr.1.gz
/usr/share/man/man1/flux-module.1.gz
/usr/share/man/man1/flux-ping.1.gz
/usr/share/man/man1/flux-ps.1.gz
/usr/share/man/man1/flux-setattr.1.gz
/usr/share/man/man1/flux-snoop.1.gz
/usr/share/man/man1/flux-start.1.gz
/usr/share/man/man1/flux-submit.1.gz
/usr/share/man/man1/flux-topo.1.gz
/usr/share/man/man1/flux-up.1.gz
/usr/share/man/man1/flux-wreck.1.gz
/usr/share/man/man1/flux-wreckrun.1.gz
/usr/share/man/man1/flux.1.gz
/usr/share/man/man3/FLUX_FATAL.3.gz
/usr/share/man/man3/flux_attr_first.3.gz
/usr/share/man/man3/flux_attr_get.3.gz
/usr/share/man/man3/flux_attr_next.3.gz
/usr/share/man/man3/flux_attr_set.3.gz
/usr/share/man/man3/flux_aux_get.3.gz
/usr/share/man/man3/flux_aux_set.3.gz
/usr/share/man/man3/flux_check_watcher_create.3.gz
/usr/share/man/man3/flux_child_watcher_create.3.gz
/usr/share/man/man3/flux_child_watcher_get_rpid.3.gz
/usr/share/man/man3/flux_child_watcher_get_rstatus.3.gz
/usr/share/man/man3/flux_close.3.gz
/usr/share/man/man3/flux_event_subscribe.3.gz
/usr/share/man/man3/flux_event_unsubscribe.3.gz
/usr/share/man/man3/flux_fatal_error.3.gz
/usr/share/man/man3/flux_fatal_set.3.gz
/usr/share/man/man3/flux_fd_watcher_create.3.gz
/usr/share/man/man3/flux_fd_watcher_get_fd.3.gz
/usr/share/man/man3/flux_flags_get.3.gz
/usr/share/man/man3/flux_flags_set.3.gz
/usr/share/man/man3/flux_flags_unset.3.gz
/usr/share/man/man3/flux_get_arity.3.gz
/usr/share/man/man3/flux_get_rank.3.gz
/usr/share/man/man3/flux_get_reactor.3.gz
/usr/share/man/man3/flux_get_size.3.gz
/usr/share/man/man3/flux_handle_watcher_create.3.gz
/usr/share/man/man3/flux_handle_watcher_get_flux.3.gz
/usr/share/man/man3/flux_idle_watcher_create.3.gz
/usr/share/man/man3/flux_msg_cmp.3.gz
/usr/share/man/man3/flux_msg_decode.3.gz
/usr/share/man/man3/flux_msg_encode.3.gz
/usr/share/man/man3/flux_msg_handler_addvec.3.gz
/usr/share/man/man3/flux_msg_handler_create.3.gz
/usr/share/man/man3/flux_msg_handler_delvec.3.gz
/usr/share/man/man3/flux_msg_handler_destroy.3.gz
/usr/share/man/man3/flux_msg_handler_start.3.gz
/usr/share/man/man3/flux_msg_handler_stop.3.gz
/usr/share/man/man3/flux_msg_recvfd.3.gz
/usr/share/man/man3/flux_msg_sendfd.3.gz
/usr/share/man/man3/flux_open.3.gz
/usr/share/man/man3/flux_pollevents.3.gz
/usr/share/man/man3/flux_pollfd.3.gz
/usr/share/man/man3/flux_prepare_watcher_create.3.gz
/usr/share/man/man3/flux_reactor_create.3.gz
/usr/share/man/man3/flux_reactor_destroy.3.gz
/usr/share/man/man3/flux_reactor_run.3.gz
/usr/share/man/man3/flux_reactor_stop.3.gz
/usr/share/man/man3/flux_reactor_stop_error.3.gz
/usr/share/man/man3/flux_recv.3.gz
/usr/share/man/man3/flux_reduce_append.3.gz
/usr/share/man/man3/flux_reduce_create.3.gz
/usr/share/man/man3/flux_reduce_destroy.3.gz
/usr/share/man/man3/flux_reduce_opt_get.3.gz
/usr/share/man/man3/flux_reduce_opt_set.3.gz
/usr/share/man/man3/flux_reduce_pop.3.gz
/usr/share/man/man3/flux_reduce_push.3.gz
/usr/share/man/man3/flux_requeue.3.gz
/usr/share/man/man3/flux_respond.3.gz
/usr/share/man/man3/flux_respond_raw.3.gz
/usr/share/man/man3/flux_rpc.3.gz
/usr/share/man/man3/flux_rpc_check.3.gz
/usr/share/man/man3/flux_rpc_completed.3.gz
/usr/share/man/man3/flux_rpc_destroy.3.gz
/usr/share/man/man3/flux_rpc_get.3.gz
/usr/share/man/man3/flux_rpc_get_raw.3.gz
/usr/share/man/man3/flux_rpc_multi.3.gz
/usr/share/man/man3/flux_rpc_raw.3.gz
/usr/share/man/man3/flux_rpc_then.3.gz
/usr/share/man/man3/flux_send.3.gz
/usr/share/man/man3/flux_set_reactor.3.gz
/usr/share/man/man3/flux_signal_watcher_create.3.gz
/usr/share/man/man3/flux_stat_watcher_create.3.gz
/usr/share/man/man3/flux_stat_watcher_get_rstat.3.gz
/usr/share/man/man3/flux_timer_watcher_create.3.gz
/usr/share/man/man3/flux_timer_watcher_reset.3.gz
/usr/share/man/man3/flux_watcher_destroy.3.gz
/usr/share/man/man3/flux_watcher_start.3.gz
/usr/share/man/man3/flux_watcher_stop.3.gz
/usr/share/man/man3/flux_zmq_watcher_create.3.gz
/usr/share/man/man3/flux_zmq_watcher_get_zsock.3.gz
/usr/share/man/man7/flux-broker-attributes.7.gz
```

**Comment by garlick** [comment]
OK, I finally got python-cffi 1.5.0 and then flux-core built as RPM's for TOSS 3.

Jimmy is going to install it on "upgrade" momentarily.

**Comment by garlick** [comment]
OK, the problem described in flux-framework/flux-core#530 and fixed in PR #526 could actually be breaking MPI launch under SLURM on our TOSS 3 test system, I'd like to propose that we tag flux-core 0.2.0 soon.  Do we have other work that this should wait for?

**Comment by lipari** [comment]
I have a quick PR that is currently building.  It will be submitted momentarily.

**Comment by grondo** [comment]
Regarding version naming, is it 0.2.0 or 0.1.1?

It would be nice to have `flux version` in there... I was working on that one, if we just report `flux.c` version it will be trivial...

**Comment by garlick** [comment]
according to semver.org

> Given a version number MAJOR.MINOR.PATCH, increment the:
> 
> MAJOR version when you make incompatible API changes,
MINOR version when you add functionality in a backwards-compatible manner, and
PATCH version when you make backwards-compatible bug fixes.
Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

[snip]

> How should I deal with revisions in the 0.y.z initial development phase?

> The simplest thing to do is start your initial development release at 0.1.0 and then increment the minor version for each subsequent release.

So I figured 0.2.0

**Comment by grondo** [comment]
I'm fine with 0.2.0 -- but PATCH isn't used in the 0.y.z phase? That might get awkward unless we get to 1.0.0 quickly.

**Comment by garlick** [comment]
Yep why don't you throw in your `flux version` that just reports PACKAGE_NAME-PACKAGE_VERSION, and we can add more to it later.

I guess if we decide to use PATCH it would not be high crime against semantic versioning :-)  However in this case I think the changes are intrusive and fail to qualify for the non-development PATCH definition.

**Comment by garlick** [comment]
To recap here, it's not possible to build flux-core against RHEL/Centos/EPEL packages, however we've updated the necessary packages in TOSS 3 and are working with Red Hat to get the prereqs updated upstream.

Since both flux-core and flux-sched are buildable on TOSS 3 now (though sched is not yet packaged and the packaged 0.2.0 core is not yet suitable to build sched against), I think this issue is resolved.  Reopen if not.
